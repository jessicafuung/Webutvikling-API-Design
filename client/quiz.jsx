import { Link, Route, Routes, useNavigate } from "react-router-dom";
import React, { useContext, useState } from "react";
import { isCorrectAnswer, randomQuestion } from "../server/question";

//Styre spørsmål fra testen
export const QuestionContext = React.createContext({ randomQuestion });

function ShowAnswer() {
  return (
    <div>
      <Routes>
        <Route path={"correct"} element={<h>Correct!</h>} />
        <Route path={"wrong"} element={<h>Wrong!</h>} />
      </Routes>

      <div>
        <Link to={"/"}>Show score</Link>
      </div>
      <div>
        <Link to={"/question"}>New question</Link>
      </div>
    </div>
  );
}

export function NewQuiz({ setQuestionAnswered, setCorrectAnswers }) {
  function handlerAnswer(answer) {
    setQuestionAnswered((q) => q + 1);
    if (isCorrectAnswer(question, answer)) {
      setCorrectAnswers((q) => q + 1);
      navigate("/answer/correct");
    } else {
      navigate("/answer/wrong");
    }
  }

  const navigate = useNavigate();
  const { randomQuestion } = useContext(QuestionContext);
  const [question] = useState(randomQuestion());

  return (
    <div>
      <h1>{question.question}</h1>
      {Object.keys(question.answers)
        .filter((a) => question.answers[a])
        .map((a) => (
          <div key={a} data-testid={a}>
            <button
              onClick={() => handlerAnswer(a)}
              style={{ fontSize: "1.5rem" }}
            >
              {question.answers[a]}
            </button>
          </div>
        ))}
    </div>
  );
}

export function FrontPage({ questionAnswered, correctAnswers }) {
  return (
    <div>
      <h1>Quiz</h1>
      <div data-testid={"status"}>
        You have answered {correctAnswers} of {questionAnswered} correctly!
      </div>
      <Link to={"/question"}>
        <button type="button" style={{ fontSize: "1.5rem" }}>
          New quiz
        </button>
      </Link>
    </div>
  );
}

export function App() {
  const [questionAnswered, setQuestionAnswered] = useState(0);
  const [correctAnswers, setCorrectAnswers] = useState(0);

  return (
    <Routes>
      <Route
        path="/"
        element={
          <FrontPage
            questionAnswered={questionAnswered}
            correctAnswers={correctAnswers}
          />
        }
      />
      <Route
        path="/question"
        element={
          <NewQuiz
            setQuestionAnswered={setQuestionAnswered}
            setCorrectAnswers={setCorrectAnswers}
          />
        }
      />
      <Route path="/answer/*" element={<ShowAnswer />} />
    </Routes>
  );
}
