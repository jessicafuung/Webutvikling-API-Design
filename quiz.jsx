import {Link, Route, Routes} from "react-router-dom";

function ShowAnswer() {
    return <h1>Answer</h1>;
}

function NewQuiz() {
    return <h1>New Quiz</h1>
}

function FrontPage() {
    return <div>
        <h1>Quiz</h1>
        <div>
            <Link to={"/question"}>
                <button type="button" style={{fontSize: "1.5rem"}}>New quiz</button>
            </Link>
            <Link to={"/answer"}>
                <button type="button" style={{fontSize: "1.5rem"}}>Answer</button>
            </Link>
        </div>
    </div>
}

export function App() {
    return (
        <Routes>
            <Route path="/" element={<FrontPage />}></Route>
            <Route path="/question" element={<NewQuiz />}></Route>
            <Route path="/answer" element={<ShowAnswer />}></Route>
        </Routes>
    );
}