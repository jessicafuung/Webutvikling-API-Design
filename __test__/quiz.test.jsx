import {App, FrontPage, NewQuiz, QuestionContext} from "../quiz";
import * as ReactDOM from "react-dom";
import {MemoryRouter} from "react-router-dom";
import React from "react";
import pretty from "pretty";
import {Simulate} from "react-dom/test-utils";

//Lage egne questions
const question = {
    question: "Are you happy?",
    answers: {
        answer_a: "Yes",
        answer_b: "No",
        answer_c: "Maybe",
    },
    correct_answers: {
        answer_a_correct: "true",
        answer_b_correct: "false",
        answer_c_correct: "false",
    }
};

describe("Quiz application",() => {

    it("shows answer status", () => {
        const element = document.createElement("div");
        ReactDOM.render(
            <MemoryRouter><FrontPage correctAnswers={3} questionAnswered={10} /></MemoryRouter>,
            element
        );
        expect(element.querySelector("[data-testid=status]").textContent)
            .toEqual("You have answered 3 of 10 correctly!")
        expect(pretty(element.innerHTML)).toMatchSnapshot();
    })

    it("shows question", () => {
        const element = document.createElement("div");

        //Komme direkte inn i /question og sende med selvlagde spørsmål
        ReactDOM.render(
            <MemoryRouter initialEntries={["/question"]}>
                <QuestionContext.Provider value={{randomQuestion: () => question}}>
                    <App />
                </QuestionContext.Provider>
            </MemoryRouter>,
            element
        );

        expect(pretty(element.innerHTML)).toMatchSnapshot();
    })

    it("records correct answer", () => {
        const setQuestionsAnswered = jest.fn();
        const setCorrectAnswers = jest.fn();

        const element = document.createElement("div");
        ReactDOM.render(
            <MemoryRouter initialEntries={["/question"]}>
                <QuestionContext.Provider value={{randomQuestion: () => question}}>
                    <NewQuiz setCorrectAnswers={setCorrectAnswers} setQuestionAnswered={setQuestionsAnswered} />
                </QuestionContext.Provider>
            </MemoryRouter>,
            element
        );

        Simulate.click(element.querySelector("[data-testid=answer_a] button"));
        expect(setQuestionsAnswered).toBeCalled();
        expect(setCorrectAnswers).toBeCalled();
        expect(pretty(element.innerHTML)).toMatchSnapshot();
    })
})