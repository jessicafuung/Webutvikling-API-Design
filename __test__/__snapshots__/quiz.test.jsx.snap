// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz application records correct answer 1`] = `
"<div>
  <h1>Are you happy?</h1>
  <div data-testid=\\"answer_a\\"><button style=\\"font-size: 1.5rem;\\">Yes</button></div>
  <div data-testid=\\"answer_b\\"><button style=\\"font-size: 1.5rem;\\">No</button></div>
  <div data-testid=\\"answer_c\\"><button style=\\"font-size: 1.5rem;\\">Maybe</button></div>
</div>"
`;

exports[`Quiz application shows answer status 1`] = `
"<div>
  <h1>Quiz</h1>
  <div data-testid=\\"status\\">You have answered 3 of 10 correctly!</div><a href=\\"/question\\"><button type=\\"button\\" style=\\"font-size: 1.5rem;\\">New quiz</button></a>
</div>"
`;

exports[`Quiz application shows question 1`] = `
"<div>
  <h1>Are you happy?</h1>
  <div data-testid=\\"answer_a\\"><button style=\\"font-size: 1.5rem;\\">Yes</button></div>
  <div data-testid=\\"answer_b\\"><button style=\\"font-size: 1.5rem;\\">No</button></div>
  <div data-testid=\\"answer_c\\"><button style=\\"font-size: 1.5rem;\\">Maybe</button></div>
</div>"
`;
